Improve image sample

MaartenBent opened the pull request:
Don't use hard-coded sizes and positions.
Make sizes DPI aware.
Fix initial size of Filter Test window.
Fix showing degree symbol in Filter Test window.

This should make it a lot easier to add new image formats to the canvas.

MaartenBent committed:
Improve image sample

Don't use hard-coded sizes and positions.
Make sizes DPI aware.

vadz commented on samples/image/image.cpp:
```diff
@@ -783,7 +783,7 @@ class MyFiltersFrame : public wxFrame
 
         wxStaticBoxSizer *sizerHue = new wxStaticBoxSizer(new wxStaticBox(this,
-                                     wxID_ANY, wxS("Hue (°)")), wxVERTICAL);
+                                     wxID_ANY, wxS("Hue (\u00b0)")), wxVERTICAL);
```
Shouldn't this `wxS` be `L`?

MaartenBent replied:
`wxS` expands to `L`, at least for me. Is there a better / preferred way?

<img src="https://github.com/user-attachments/assets/9dd3b4ed-456c-4ed6-ab7d-4a7063d3d609" />

vadz replied:
It expands to nothing in UTF-8 build, which is the whole reason for having it in addition to wxT — which does always expand to `L`.

So there are several alternative ways to write this, none of which is ideal:

1. `wxT("\u00b0")`
2. `L"\u00b0"`
3. `wxString::FromUTF8("\xc2\xb0")`

Ideal would be to write either `L"°"` or `wxString::FromUTF8("°")` but this would require using UTF-8 encoding for the sources under all platforms and I'm not sure if we do this (in fact, I'm almost sure that we don't as we don't use `/utf-8` with MSVC).

vadz commented on samples/image/canvas.cpp:
```diff
@@ -44,27 +34,23 @@ wxBEGIN_EVENT_TABLE(MyCanvas, wxScrolledWindow)
 
-    wxBitmap bitmap( 100, 100 );
+    wxBitmap bitmap;
+    bitmap.CreateWithLogicalSize(FromDIP(wxSize(100, 100)), GetDPIScaleFactor());
```
Wouldn't `CreateWithDIPSize()` work here?

MaartenBent replied:
I reverted these changes and always create a bitmap with scale 1. This way bitmap loaded from file and the bitmap created for the canvas will always have the same scale, even after DPI events.

vadz commented on samples/image/canvas.cpp:
```diff
@@ -396,266 +384,348 @@ MyCanvas::~MyCanvas()
+{
+    // Aproximate the size used in OnPaint.
+    const int width = 4 * FromDIP(250);
+    const int height = 3100;
```
Strange to use `FromDIP()` for width but not for height, is this intentional?

vadz commented on samples/image/image.cpp:
```diff
@@ -1161,11 +1161,13 @@ MyFrame::MyFrame()
-    m_canvas->SetScrollbars( 10, 10, 50, 275 );
+    m_canvas = new MyCanvas(this);
+    wxSize cz = m_canvas->GetDrawingSize();
+    m_canvas->SetScrollbars(10, 10, cz.GetWidth() / 10, cz.GetHeight() / 10);
```
No need for `FromDIP()` here?

MaartenBent replied:
No, the `SetScrollbars` arguments (`pixelsPerUnit` * `noUnits`) should match the number of pixels of the canvas.

It is actually not per-monitor DPI aware, so I'll add a DPI event handler and update `SetScrollbars` when the canvas size changes.

vadz reviewed (commented):
Thanks, looks good, but I have a couple of minor questions, please let me know if you're going to change anything or if you think this should be merged as is.

MaartenBent committed:
Improve image filter window in image sample

Fix showing the degrees icon.
Set the initial size of the window.

MaartenBent committed:
Show full library versions in image sample

MaartenBent committed:
Add DPI-scaled image to image sample

All the default images are drawn at their logical size.
For systems that do not have DPI independent pixels (like Windows or gtk2),
show an image that is scaled to the active DPI.

vadz commented:
Maarten, should this be merged now?

MaartenBent commented:
Yes please. I don't have any more changes.
